<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Dandelions: TheApp Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Dandelions
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="class_the_app-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">TheApp Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<div class="dynheader">
Inheritance diagram for TheApp:</div>
<div class="dyncontent">
 <div class="center">
  <img src="class_the_app.png" alt=""/>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:acde0d598ea2d68f16b1df9f7c6b82951"><td class="memItemLeft" align="right" valign="top"><a id="acde0d598ea2d68f16b1df9f7c6b82951"></a>
virtual bool&#160;</td><td class="memItemRight" valign="bottom"><b>OnInit</b> ()</td></tr>
<tr class="separator:acde0d598ea2d68f16b1df9f7c6b82951"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4fc9037b5415169fc3a8ec2e27bc7698"><td class="memItemLeft" align="right" valign="top"><a id="a4fc9037b5415169fc3a8ec2e27bc7698"></a>
virtual bool&#160;</td><td class="memItemRight" valign="bottom"><b>OnExceptionInMainLoop</b> ()</td></tr>
<tr class="separator:a4fc9037b5415169fc3a8ec2e27bc7698"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1999176ddc9afb206687345d92075d72"><td class="memItemLeft" align="right" valign="top"><a id="a1999176ddc9afb206687345d92075d72"></a>
virtual void&#160;</td><td class="memItemRight" valign="bottom"><b>OnUnhandledException</b> ()</td></tr>
<tr class="separator:a1999176ddc9afb206687345d92075d72"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Build and display a consensus minimum spanning tree (MST) that (hopefully) reflects phylogeny and clustering of a set of related nucleotide sequences.<br  />
 Instructions for use:<br  />
 From the main window, File-&gt;Open one of the following: </p><ol>
<li>
a .dsa output file (run with &ndash;template_dna=... and &ndash;show_codons=horizontal options) </li>
<li>
or a .fasta file of nucleotide sequences </li>
<li>
or a or plain .txt file with a nucleotide sequence on each line </li>
</ol>
<p>IMPORTANT: when opening dsa output, the tree will be rooted on the &ndash;template_dna sequence. For a .fasta or .txt file, the FIRST sequence encountered in the file will be used as the root. The program will remove duplicate sequences and any sequence whose length differs from that of the root (i.e., it does not yet handle indels).<br  />
 When prompted, enter the number of MSTs on which to base the consensus. The consensus algorith works such that for a sample size of 1 (the default) the 'consensus' is the same as the single input tree. Details of the consensus building algorith are give at the end. The nucleotide consensus tree is then transformed into a tree of amino acid sequencs by merging all nodes in each subtree of the tree that share the same translation. The size of the nodes when displayed is proportional to the number of nodes merged in this process. Nodes with unusually high numbers (&gt;2 s.d. above the mean) of direct descendants are designated as 'centroids' and given special treatment. Each centroid is assigned a color and a number starting from 1. The lower the number, the more descendants a given centroid has. Centroids and the nodes and edges that connect them to the root node adopt the color of their centroid. The root node is painted black.<br  />
 The Nodes and edges are displayed using an interactive physics simulation similar to that of the vis.js library. Nodes are modeled as point masses that repel each other. Edges are modeled as springs. <a class="el" href="struct_edge.html">Edge</a> length is proportional to the sequence distance between two nodes.<br  />
 The following parameters goven the simulation: </p><ol>
<li>
repulsion - 1/r^2 repulsive force between Nodes </li>
<li>
tension - spring constant </li>
<li>
scale - scale factor applied to the length of each edge/spring </li>
<li>
drag - drag coefficient </li>
<li>
compaction - global force field that pushes nodes toward the origin </li>
<li>
stability - limit to the maximum node velocity </li>
<li>
time - length of the timestep for each iteration </li>
</ol>
<p>The graph can be exported to .svg or .png format using File-&gt;Export-&gt;Graphic. File-&gt;Export-&gt;Mutations produces a formatted and color-coded table of sequence differences between each centroid and the root in a .html file.<br  />
 The MST and consensus building algorithms:<br  />
 The individual MSTs that make up the consensus tree are built by first randomly shuffling the input sequences and following Prim's algorithm with the following modification: distance between nodes p (aleady in tree) and c (not yet in tree) is a 2-tuple of (D(c,p), D(p,r)) where r is the root node and D(i,j) is the Hamming Distance between the two seqeunces. This change generates trees that reflect a breadth-first approach search of the underlying biological system (as we expect for affinity maturation).<br  />
 To build the conensus tree, we generate n MSTs, each from a random permutation of the input sequences. Optionally, phylogenetic inference is performed on the input sequences with each iteration and the inferred ancestors are incorporated during MST construction, then removed from the final tree. We keep track of the frequency with which every node (except root) c is the direct descendant of every other node p. This results in a (k-1)x(k) element distance matrix (where k is the number of nodes and root has no ancestor) with elements M(c,p) in the range [0, 1]. We then construct a single MST from M with edge distances of 1-M[c,p]. The algorithm has the nice property that for a sample size of n=1, the output consensus tree is identical to its single n=1 input tree. </p>
</div><hr/>The documentation for this class was generated from the following file:<ul>
<li>src/main.cpp</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
